### ISSUE TITLE
reproducible crash on update(), large(ish) dataset

### ISSUE THREAD

**Issue by silver83** [rethinkdb/rethinkdb#4701]
Seems like { returnChanges: 'always' } is making it crash faster.... 
Ran on a smaller dataset and passed without issue.
Ran again with { returnChange: false } and it just took much longer but it crashed after a 3 minutes or so...

Seems like its growing from 400MB before the query, to ~5GB.

Also noticed that when it doesn't crash, but rather just times out in the data explorer, the memory footprint doesn't return to pre-query levels...

Table1 count = 2018451
Table2 count = 1010070

```
db.table('Table1').update(function(link) {
    var table2 = db.table('Table2').get(link('Table2_id'));
	  return { updatedAt: table2('updatedAt') };
}, { returnChanges: 'always', nonAtomic:true });
```

```
Running rethinkdb 2.0.1 (CLANG 6.1.0 (clang-602.0.49))...
Running on Darwin 14.4.0 x86_64
Loading data from directory /Users/yoni.shalom/rethinkdb_data
Listening for intracluster connections on port 29015
Listening for client driver connections on port 28015
Listening for administrative HTTP connections on port 8080
Listening on addresses: [redacted]
Server ready, "redacted" 5c7677f4-575d-4850-9935-76186caab49d
A newer version of the RethinkDB server is available: 2.0.4. You can read the changelog at <https://github.com/rethinkdb/rethinkdb/releases>.
Version: rethinkdb 2.0.1 (CLANG 6.1.0 (clang-602.0.49))
error: Error in src/rdb_protocol/profile.cc at line 241:
error: Guarantee failed: [stop]
error: Backtrace:
error: Mon Aug 17 14:33:17 2015

       1: 0   rethinkdb                           0x0000000108e375e3 _Z19rethinkdb_backtracePPvi + 132 at 0x108e375e3 ()
       2: 0   rethinkdb                           0x0000000108acc51c _ZN11backtrace_tC2Ev + 72 at 0x108acc51c ()
       3: 0   rethinkdb                           0x0000000108accc6d _ZN26lazy_backtrace_formatter_tC2Ev + 39 at 0x108accc6d ()
       4: 0   rethinkdb                           0x0000000108acc1b5 _Z16format_backtraceb + 25 at 0x108acc1b5 ()
       5: 0   rethinkdb                           0x0000000108c61295 _Z18report_fatal_errorPKciS0_z + 412 at 0x108c61295 ()
       6: 0   rethinkdb                           0x0000000108d57aa3 _ZN7profile8durationERKNSt3__16vectorIN5boost7variantINS_7start_tENS_7split_tENS_8sample_tENS_6stop_tENS2_6detail7variant5void_ESA_SA_SA_SA_SA_SA_SA_SA_SA_SA_SA_SA_SA_SA_SA_EENS0_9allocatorISB_EEEE + 213 at 0x108d57aa3 ()
       7: 0   rethinkdb                           0x0000000108d57af5 _ZN7profile9sampler_t10new_sampleEv + 33 at 0x108d57af5 ()
       8: 0   rethinkdb                           0x0000000108ca6da0 _Z33do_a_replace_from_batched_replaceN14auto_drainer_t6lock_tEP20fifo_enforcer_sink_tRK27fifo_enforcer_write_token_tRK16btree_loc_info_t13one_replace_tRKN2ql19configured_limits_tEP9promise_tIP12superblock_tEP28rdb_modification_report_cb_tbPNSA_7datum_tEPN7profile9sampler_tEPNSN_7trace_tEPNSt3__13setINSS_12basic_stringIcNSS_11char_traitsIcEENSS_9allocatorIcEEEENSS_4lessISZ_EENSX_ISZ_EEEE + 58 at 0x108ca6da0 ()
       9: 0   rethinkdb                           0x0000000108cb07b6 _ZNSt3__110__function6__funcINS_6__bindIPFvN14auto_drainer_t6lock_tEP20fifo_enforcer_sink_tRK27fifo_enforcer_write_token_tRK16btree_loc_info_t13one_replace_tRKN2ql19configured_limits_tEP9promise_tIP12superblock_tEP28rdb_modification_report_cb_tbPNSE_7datum_tEPN7profile9sampler_tEPNSR_7trace_tEPNS_3setINS_12basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEENS_4lessIS12_EENS10_IS12_EEEEEJS4_S6_S7_SA_SD_RSF_SM_RSO_RbSQ_RST_RSV_S17_EEENS10_IS1F_EEFvvEEclEv + 172 at 0x108cb07b6 ()
       10: 0   rethinkdb                           0x0000000108bf0f07 _ZN11coro_pool_tINSt3__18functionIFvvEEEE10worker_runES3_N14auto_drainer_t6lock_tE + 135 at 0x108bf0f07 ()
       11: 0   rethinkdb                           0x0000000108c4ac13 _ZN26callable_action_instance_tINSt3__16__bindIM11coro_pool_tINS0_8functionIFvvEEEEFvS5_N14auto_drainer_t6lock_tEEJPS6_S5_S8_EEEE10run_actionEv + 107 at 0x108c4ac13 ()
       12: 0   rethinkdb                           0x0000000108ac4a02 _ZN6coro_t3runEv + 70 at 0x108ac4a02 ()
error: Exiting.
[1]    41562 trace trap  rethinkdb --bind all
```

**Comment by 1lann** [comment]
Seemed to have been fixed in 2.0.4: #4430

**Comment by danielmewes** [comment]
Yeah this looks like a duplicate of https://github.com/rethinkdb/rethinkdb/issues/4430 like @1lann says.

@silver83 can you install RethinkDB 2.0.4? It should fix the problem.
